Please add support for pydantic-2

The package currently requires pydantic-1. Please add support for pydantic-2. FWICS it would require some code changes:

```pytb
ImportError while loading conftest '/tmp/wtfis/tests/conftest.py'.
tests/conftest.py:8: in <module>
    from wtfis.models.greynoise import GreynoiseIp
wtfis/models/greynoise.py:18: in <module>
    class GreynoiseIpMap(BaseModel):
.venv/lib/python3.11/site-packages/pydantic/_internal/_model_construction.py:95: in __new__
    private_attributes = inspect_namespace(
.venv/lib/python3.11/site-packages/pydantic/_internal/_model_construction.py:279: in inspect_namespace
    raise TypeError("To define root models, use `pydantic.RootModel` rather than a field called '__root__'")
E   TypeError: To define root models, use `pydantic.RootModel` rather than a field called '__root__'
```

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -27,7 +27,7 @@ jobs:
       fail-fast: false
       matrix:
         os: [ubuntu-latest, windows-latest, macos-latest]
-        python-version: ['3.8', '3.9', '3.10']
+        python-version: ['3.8', '3.9', '3.10', '3.11']
 
     steps:
     - uses: actions/checkout@v3

--- a/.gitignore
+++ b/.gitignore
@@ -162,5 +162,8 @@ cython_debug/
 # VSCode
 .vscode/
 
+# hatch virtualenvs
+.hatch/
+
 # Custom env
 .env.wtfis

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -78,7 +78,7 @@ test-all = [
 
 # Hatch test env
 [[tool.hatch.envs.test.matrix]]
-python = ["38", "39", "310"]
+python = ["38", "39", "310", "311"]
 
 # Hatch lint env
 [tool.hatch.envs.lint]

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -59,6 +59,8 @@ disable = true
 # Hatch default env
 [tool.hatch.envs.default]
 dependencies = [
+    "bandit",
+    "flake8>=6.0.0",
     "freezegun",
     "mypy",
     "pytest",
@@ -81,7 +83,7 @@ python = ["38", "39", "310"]
 # Hatch lint env
 [tool.hatch.envs.lint]
 detached = true
-dependencies = [
+dependencies = [  # Make sure the respective versions are synced with default!
     "bandit",
     "flake8>=6.0.0",
 ]

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -29,7 +29,7 @@ classifiers = [
     "Topic :: Security",
 ]
 dependencies = [
-    "pydantic~=1.10.2",
+    "pydantic~=2.0.3",
     "python-dotenv~=0.21.0",
     "requests~=2.28.1",
     "rich~=12.6.0",

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -35,17 +35,17 @@ def open_test_data(fname: str) -> str:
 
 def greynoise_get(ip, pool) -> GreynoiseIp:
     """ Mock replacement for GreynoiseClient().get_ip() """
-    return GreynoiseIp.parse_obj(pool[ip])
+    return GreynoiseIp.model_validate(pool[ip])
 
 
 def ipwhois_get(ip, pool) -> IpWhois:
     """ Mock replacement for IpWhoisClient().get_ipwhois() """
-    return IpWhois.parse_obj(pool[ip])
+    return IpWhois.model_validate(pool[ip])
 
 
 def shodan_get_ip(ip, pool) -> ShodanIp:
     """ Mock replacement for ShodanClient().get_ip() """
-    return ShodanIp.parse_obj(pool[ip])
+    return ShodanIp.model_validate(pool[ip])
 
 
 def timestamp_text(ts) -> Optional[RenderableType]:

--- a/tests/test_cli.py
+++ b/tests/test_cli.py
@@ -444,7 +444,7 @@ class TestGenView:
             whois_client=MagicMock(),
             greynoise_client=MagicMock(),
         )
-        entity.vt_info = Domain.parse_obj(json.loads(test_data("vt_domain_gist.json")))
+        entity.vt_info = Domain.model_validate(json.loads(test_data("vt_domain_gist.json")))
         entity.whois = MagicMock()
         entity.ip_enrich = MagicMock()
         view = generate_view(MagicMock(), MagicMock(), entity)
@@ -462,7 +462,7 @@ class TestGenView:
             whois_client=MagicMock(),
             greynoise_client=MagicMock(),
         )
-        entity.vt_info = IpAddress.parse_obj(json.loads(test_data("vt_ip_1.1.1.1.json")))
+        entity.vt_info = IpAddress.model_validate(json.loads(test_data("vt_ip_1.1.1.1.json")))
         entity.whois = MagicMock()
         entity.ip_enrich = MagicMock()
         view = generate_view(MagicMock(), MagicMock(), entity)

--- a/tests/test_handlers.py
+++ b/tests/test_handlers.py
@@ -57,7 +57,7 @@ class TestDomainHandler:
     def test_fetch_data_1(self, domain_handler, test_data):
         """ Test with max_resolutions = 3 (default) """
         handler = domain_handler()
-        handler.resolutions = Resolutions.parse_obj(json.loads(test_data("vt_resolutions_gist.json")))
+        handler.resolutions = Resolutions.model_validate(json.loads(test_data("vt_resolutions_gist.json")))
 
         handler._fetch_vt_domain = MagicMock()
         handler._fetch_vt_resolutions = MagicMock()
@@ -135,9 +135,9 @@ class TestDomainHandler:
 
             capture = capsys.readouterr()
 
-            assert capture.err == (
+            assert capture.err.startswith(
                 "Data model validation error: 1 validation error for Domain\ndata\n"
-                "  field required (type=value_error.missing)\n"
+                "  Field required [type=missing, input_value={'intentionally': 'wrong data'}, input_type=dict]\n"
             )
             assert e.type == SystemExit
             assert e.value.code == 1
@@ -150,7 +150,7 @@ class TestDomainHandler:
     @patch.object(requests.Session, "get")
     def test_ipwhois_http_error(self, mock_requests_get, domain_handler, capsys, test_data):
         handler = domain_handler()
-        handler.resolutions = Resolutions.parse_obj(json.loads(test_data("vt_resolutions_gist.json")))
+        handler.resolutions = Resolutions.model_validate(json.loads(test_data("vt_resolutions_gist.json")))
 
         mock_resp = requests.models.Response()
 
@@ -231,7 +231,7 @@ class TestDomainHandler:
         mock_resp.status_code = 429
         mock_requests_get.return_value = mock_resp
 
-        handler.resolutions = Resolutions.parse_obj(json.loads(test_data("vt_resolutions_gist.json")))
+        handler.resolutions = Resolutions.model_validate(json.loads(test_data("vt_resolutions_gist.json")))
 
         handler._fetch_greynoise()
         assert handler.warnings[0].startswith("Could not fetch Greynoise: 429 Client Error:")
@@ -248,14 +248,14 @@ class TestDomainHandler:
         handler = domain_handler(3)
         mock_resp = requests.models.Response()
 
-        handler.resolutions = Resolutions.parse_obj(json.loads(test_data("vt_resolutions_gist.json")))
+        handler.resolutions = Resolutions.model_validate(json.loads(test_data("vt_resolutions_gist.json")))
 
         mock_resp.status_code = 404
         mock_requests_get.return_value = mock_resp
 
         handler._fetch_greynoise()
         assert len(handler.warnings) == 0
-        assert handler.greynoise == GreynoiseIpMap(__root__={})
+        assert handler.greynoise == GreynoiseIpMap.model_validate({})
 
 
 class TestIpAddressHandler:
@@ -310,9 +310,9 @@ class TestIpAddressHandler:
 
             capture = capsys.readouterr()
 
-            assert capture.err == (
+            assert capture.err.startswith(
                 "Data model validation error: 1 validation error for IpAddress\ndata\n"
-                "  field required (type=value_error.missing)\n"
+                "  Field required [type=missing, input_value={'intentionally': 'wrong data'}, input_type=dict]\n"
             )
             assert e.type == SystemExit
             assert e.value.code == 1
@@ -403,4 +403,4 @@ class TestIpAddressHandler:
 
         handler._fetch_greynoise()
         assert len(handler.warnings) == 0
-        assert handler.greynoise == GreynoiseIpMap(__root__={})
+        assert handler.greynoise == GreynoiseIpMap.model_validate({})

--- a/tests/test_models_vt.py
+++ b/tests/test_models_vt.py
@@ -7,7 +7,7 @@ from wtfis.models.virustotal import (
 
 class TestVirustotalModels:
     def test_domain_1(self, test_data):
-        domain = Domain.parse_obj(json.loads(test_data("vt_domain_gist.json")))
+        domain = Domain.model_validate(json.loads(test_data("vt_domain_gist.json")))
 
         assert domain.data.id_ == "gist.github.com"
         assert domain.data.type_ == "domain"
@@ -15,13 +15,13 @@ class TestVirustotalModels:
         assert domain.data.attributes.last_analysis_stats.suspicious == 0
         assert domain.data.attributes.last_analysis_stats.harmless == 84
         assert domain.data.attributes.last_analysis_stats.undetected == 10
-        assert list(domain.data.attributes.last_analysis_results.__root__.keys())[:2] == [
+        assert list(domain.data.attributes.last_analysis_results.root.keys())[:2] == [
             "CMC Threat Intelligence",
             "Snort IP sample list",
         ]
 
     def test_resolutions_1(self, test_data):
-        res = Resolutions.parse_obj(json.loads(test_data("vt_resolutions_gist.json")))
+        res = Resolutions.model_validate(json.loads(test_data("vt_resolutions_gist.json")))
 
         assert res.meta.count == 37
         assert len(res.data) == 10

--- a/tests/test_ui_domain_view.py
+++ b/tests/test_ui_domain_view.py
@@ -25,7 +25,7 @@ from wtfis.ui.view import DomainView
 @pytest.fixture()
 def view01(test_data, mock_ipwhois_get):
     """ gist.github.com with PT whois. Complete test of all panels. Also test print(). """
-    resolutions = Resolutions.parse_obj(json.loads(test_data("vt_resolutions_gist.json")))
+    resolutions = Resolutions.model_validate(json.loads(test_data("vt_resolutions_gist.json")))
 
     ipwhois_pool = json.loads(test_data("ipwhois_gist.json"))
     ipwhois_client = IpWhoisClient()
@@ -34,11 +34,11 @@ def view01(test_data, mock_ipwhois_get):
 
     return DomainView(
         console=Console(),
-        entity=Domain.parse_obj(json.loads(test_data("vt_domain_gist.json"))),
+        entity=Domain.model_validate(json.loads(test_data("vt_domain_gist.json"))),
         resolutions=resolutions,
-        whois=PTWhois.parse_obj(json.loads(test_data("pt_whois_gist.json"))),
+        whois=PTWhois.model_validate(json.loads(test_data("pt_whois_gist.json"))),
         ip_enrich=ip_enrich,
-        greynoise=GreynoiseIpMap(__root__={}),
+        greynoise=GreynoiseIpMap.model_validate({}),
     )
 
 
@@ -51,10 +51,10 @@ def view02(test_data):
     return DomainView(
         console=Console(),
         entity=MagicMock(),
-        resolutions=Resolutions.parse_obj(json.loads(test_data("vt_resolutions_gist.json"))),
-        whois=VTWhois.parse_obj(json.loads(test_data("vt_whois_gist.json"))),
-        ip_enrich=IpWhoisMap(__root__={}),
-        greynoise=GreynoiseIpMap(__root__={}),
+        resolutions=Resolutions.model_validate(json.loads(test_data("vt_resolutions_gist.json"))),
+        whois=VTWhois.model_validate(json.loads(test_data("vt_whois_gist.json"))),
+        ip_enrich=IpWhoisMap.model_validate({}),
+        greynoise=GreynoiseIpMap.model_validate({}),
         max_resolutions=1,
     )
 
@@ -66,7 +66,7 @@ def view03(test_data):
         console=Console(),
         entity=MagicMock(),
         resolutions=MagicMock(),
-        whois=VTWhois.parse_obj(json.loads(test_data("vt_whois_bbc.json"))),
+        whois=VTWhois.model_validate(json.loads(test_data("vt_whois_bbc.json"))),
         ip_enrich=MagicMock(),
         greynoise=MagicMock(),
     )
@@ -80,7 +80,7 @@ def view04(test_data):
     """
     return DomainView(
         console=Console(),
-        entity=Domain.parse_obj(json.loads(test_data("vt_domain_google.json"))),
+        entity=Domain.model_validate(json.loads(test_data("vt_domain_google.json"))),
         resolutions=None,
         whois=MagicMock(),
         ip_enrich=MagicMock(),
@@ -93,7 +93,7 @@ def view05(test_data):
     """ tucows.com domain. Domain test only. Test domain with negative reputation and no popularity."""
     return DomainView(
         console=Console(),
-        entity=Domain.parse_obj(json.loads(test_data("vt_domain_tucows.json"))),
+        entity=Domain.model_validate(json.loads(test_data("vt_domain_tucows.json"))),
         resolutions=MagicMock(),
         whois=MagicMock(),
         ip_enrich=MagicMock(),
@@ -108,7 +108,7 @@ def view06(test_data):
         console=Console(),
         entity=MagicMock(),
         resolutions=MagicMock(),
-        whois=VTWhois.parse_obj(json.loads(test_data("vt_whois_example_2.json"))),
+        whois=VTWhois.model_validate(json.loads(test_data("vt_whois_example_2.json"))),
         ip_enrich=MagicMock(),
         greynoise=MagicMock(),
     )
@@ -117,7 +117,7 @@ def view06(test_data):
 @pytest.fixture()
 def view07(test_data, mock_shodan_get_ip):
     """ gist.github.com with Shodan. Only test resolution and IP enrich. """
-    resolutions = Resolutions.parse_obj(json.loads(test_data("vt_resolutions_gist.json")))
+    resolutions = Resolutions.model_validate(json.loads(test_data("vt_resolutions_gist.json")))
 
     shodan_pool = json.loads(test_data("shodan_gist.json"))
     shodan_client = ShodanClient(MagicMock())
@@ -130,14 +130,14 @@ def view07(test_data, mock_shodan_get_ip):
         resolutions=resolutions,
         whois=MagicMock(),
         ip_enrich=ip_enrich,
-        greynoise=GreynoiseIpMap(__root__={}),
+        greynoise=GreynoiseIpMap.model_validate({}),
     )
 
 
 @pytest.fixture()
 def view08(test_data, mock_shodan_get_ip):
     """ www.wired.com with Shodan. Only test resolution and IP enrich. """
-    resolutions = Resolutions.parse_obj(json.loads(test_data("vt_resolutions_wired.json")))
+    resolutions = Resolutions.model_validate(json.loads(test_data("vt_resolutions_wired.json")))
 
     shodan_pool = json.loads(test_data("shodan_wired.json"))
     shodan_client = ShodanClient(MagicMock())
@@ -150,7 +150,7 @@ def view08(test_data, mock_shodan_get_ip):
         resolutions=resolutions,
         whois=MagicMock(),
         ip_enrich=ip_enrich,
-        greynoise=GreynoiseIpMap(__root__={}),
+        greynoise=GreynoiseIpMap.model_validate({}),
         max_resolutions=1,
     )
 
@@ -158,7 +158,7 @@ def view08(test_data, mock_shodan_get_ip):
 @pytest.fixture()
 def view09(test_data, mock_shodan_get_ip, mock_greynoise_get):
     """ one.one.one.one with Shodan. Only test resolution and IP enrich. """
-    resolutions = Resolutions.parse_obj(json.loads(test_data("vt_resolutions_one.json")))
+    resolutions = Resolutions.model_validate(json.loads(test_data("vt_resolutions_one.json")))
 
     shodan_pool = json.loads(test_data("shodan_one.json"))
     shodan_client = ShodanClient(MagicMock())
@@ -188,16 +188,16 @@ def view10(test_data):
         console=Console(),
         entity=MagicMock(),
         resolutions=MagicMock(),
-        whois=VTWhois.parse_obj(json.loads(test_data("vt_whois_foo.json"))),
+        whois=VTWhois.model_validate(json.loads(test_data("vt_whois_foo.json"))),
         ip_enrich=MagicMock(),
-        greynoise=GreynoiseIpMap(__root__={}),
+        greynoise=GreynoiseIpMap.model_validate({}),
     )
 
 
 @pytest.fixture()
 def view11(test_data, mock_shodan_get_ip):
     """ gist.github.com with Shodan. Only test IP enrich. Test empty open ports. """
-    resolutions = Resolutions.parse_obj(json.loads(test_data("vt_resolutions_gist.json")))
+    resolutions = Resolutions.model_validate(json.loads(test_data("vt_resolutions_gist.json")))
 
     shodan_pool = json.loads(test_data("shodan_gist_2.json"))
     shodan_client = ShodanClient(MagicMock())
@@ -210,7 +210,7 @@ def view11(test_data, mock_shodan_get_ip):
         resolutions=resolutions,
         whois=MagicMock(),
         ip_enrich=ip_enrich,
-        greynoise=GreynoiseIpMap(__root__={}),
+        greynoise=GreynoiseIpMap.model_validate({}),
     )
 
 
@@ -221,7 +221,7 @@ def view12(test_data):
         console=Console(),
         entity=MagicMock(),
         resolutions=MagicMock(),
-        whois=Ip2Whois.parse_obj(json.loads(test_data("ip2whois_whois_hotmail.json"))),
+        whois=Ip2Whois.model_validate(json.loads(test_data("ip2whois_whois_hotmail.json"))),
         ip_enrich=MagicMock(),
         greynoise=MagicMock(),
     )
@@ -234,7 +234,7 @@ def view13(test_data):
         console=Console(),
         entity=MagicMock(),
         resolutions=MagicMock(),
-        whois=Ip2Whois.parse_obj(json.loads(test_data("ip2whois_whois_bbc.json"))),
+        whois=Ip2Whois.model_validate(json.loads(test_data("ip2whois_whois_bbc.json"))),
         ip_enrich=MagicMock(),
         greynoise=MagicMock(),
     )

--- a/tests/test_ui_ip_view.py
+++ b/tests/test_ui_ip_view.py
@@ -36,8 +36,8 @@ def view01(test_data, mock_ipwhois_get, mock_greynoise_get):
 
     return IpAddressView(
         console=Console(),
-        entity=IpAddress.parse_obj(json.loads(test_data("vt_ip_1.1.1.1.json"))),
-        whois=PTWhois.parse_obj(json.loads(test_data("pt_whois_1.1.1.1.json"))),
+        entity=IpAddress.model_validate(json.loads(test_data("vt_ip_1.1.1.1.json"))),
+        whois=PTWhois.model_validate(json.loads(test_data("pt_whois_1.1.1.1.json"))),
         ip_enrich=ip_enrich,
         greynoise=greynoise_enrich,
     )
@@ -59,7 +59,7 @@ def view02(test_data, mock_shodan_get_ip, mock_greynoise_get):
 
     return IpAddressView(
         console=Console(),
-        entity=IpAddress.parse_obj(json.loads(test_data("vt_ip_1.1.1.1.json"))),
+        entity=IpAddress.model_validate(json.loads(test_data("vt_ip_1.1.1.1.json"))),
         whois=MagicMock(),
         ip_enrich=ip_enrich,
         greynoise=greynoise_enrich,
@@ -72,7 +72,7 @@ def view03(test_data):
     return IpAddressView(
         console=Console(),
         entity=MagicMock(),
-        whois=VTWhois.parse_obj(json.loads(test_data("vt_whois_1.1.1.1.json"))),
+        whois=VTWhois.model_validate(json.loads(test_data("vt_whois_1.1.1.1.json"))),
         ip_enrich=MagicMock(),
         greynoise=MagicMock(),
     )
@@ -86,10 +86,10 @@ def view04(test_data):
     """
     return IpAddressView(
         console=Console(),
-        entity=IpAddress.parse_obj(json.loads(test_data("vt_ip_142.251.220.110.json"))),
+        entity=IpAddress.model_validate(json.loads(test_data("vt_ip_142.251.220.110.json"))),
         whois=MagicMock(),
-        ip_enrich=IpWhoisMap(__root__={}),
-        greynoise=GreynoiseIpMap(__root__={}),
+        ip_enrich=IpWhoisMap.model_validate({}),
+        greynoise=GreynoiseIpMap.model_validate({}),
     )
 
 
@@ -104,9 +104,9 @@ def view05(test_data, mock_greynoise_get):
 
     return IpAddressView(
         console=Console(),
-        entity=IpAddress.parse_obj(json.loads(test_data("vt_ip_1.1.1.1.json"))),
+        entity=IpAddress.model_validate(json.loads(test_data("vt_ip_1.1.1.1.json"))),
         whois=MagicMock(),
-        ip_enrich=IpWhoisMap(__root__={}),
+        ip_enrich=IpWhoisMap.model_validate({}),
         greynoise=greynoise_enrich,
     )
 
@@ -122,9 +122,9 @@ def view06(test_data, mock_greynoise_get):
 
     return IpAddressView(
         console=Console(),
-        entity=IpAddress.parse_obj(json.loads(test_data("vt_ip_1.1.1.1.json"))),
+        entity=IpAddress.model_validate(json.loads(test_data("vt_ip_1.1.1.1.json"))),
         whois=MagicMock(),
-        ip_enrich=IpWhoisMap(__root__={}),
+        ip_enrich=IpWhoisMap.model_validate({}),
         greynoise=greynoise_enrich,
     )
 

--- a/wtfis/clients/greynoise.py
+++ b/wtfis/clients/greynoise.py
@@ -18,7 +18,7 @@ class GreynoiseClient(BaseClient):
         return "Greynoise"
 
     def get_ip(self, ip: str) -> GreynoiseIp:
-        return GreynoiseIp.parse_obj(
+        return GreynoiseIp.model_validate(
             self._get(f"/{ip}", headers={"key": self.api_key})
         )
 
@@ -34,11 +34,11 @@ class GreynoiseClient(BaseClient):
             ip_data = self.get_ip(ip.attributes.ip_address)
             if ip_data:
                 greynoise_map[ip_data.ip] = ip_data
-        return GreynoiseIpMap(__root__=greynoise_map)
+        return GreynoiseIpMap.model_validate(greynoise_map)
 
     def single_get_ip(self, ip: str) -> GreynoiseIpMap:
         greynoise_map = {}
         ip_data = self.get_ip(ip)
         if ip_data:
             greynoise_map[ip] = ip_data
-        return GreynoiseIpMap(__root__=greynoise_map)
+        return GreynoiseIpMap.model_validate(greynoise_map)

--- a/wtfis/clients/ip2whois.py
+++ b/wtfis/clients/ip2whois.py
@@ -27,12 +27,12 @@ class Ip2WhoisClient(BaseClient):
 
         # Let a 404 or invalid domain pass
         try:
-            return Whois.parse_obj(self._get("/", params))
+            return Whois.model_validate(self._get("/", params))
         except HTTPError as e:
             if (
                 e.response.status_code == 404 or
                 (e.response.status_code == 400 and
                  e.response.json().get("error", {})["error_code"] == 10007)
             ):
-                return Whois.parse_obj({})
+                return Whois.model_validate({})
             raise

--- a/wtfis/clients/ipwhois.py
+++ b/wtfis/clients/ipwhois.py
@@ -13,7 +13,7 @@ class IpWhoisClient(BaseClient):
 
     def get_ipwhois(self, ip: str) -> Optional[IpWhois]:
         result = self._get(f"/{ip}")
-        return IpWhois.parse_obj(result) if result.get("success") is True else None
+        return IpWhois.model_validate(result) if result.get("success") is True else None
 
     @property
     def name(self) -> str:
@@ -31,11 +31,11 @@ class IpWhoisClient(BaseClient):
             ipwhois = self.get_ipwhois(ip.attributes.ip_address)
             if ipwhois:
                 ipwhois_map[ipwhois.ip] = ipwhois
-        return IpWhoisMap(__root__=ipwhois_map)
+        return IpWhoisMap.model_validate(ipwhois_map)
 
     def single_get_ip(self, ip: str) -> IpWhoisMap:
         ipwhois_map = {}
         ipwhois = self.get_ipwhois(ip)
         if ipwhois:
             ipwhois_map[ip] = ipwhois
-        return IpWhoisMap(__root__=ipwhois_map)
+        return IpWhoisMap.model_validate(ipwhois_map)

--- a/wtfis/clients/passivetotal.py
+++ b/wtfis/clients/passivetotal.py
@@ -27,4 +27,4 @@ class PTClient(BaseClient):
         return self._query("/dns/passive", refang(domain))
 
     def get_whois(self, entity: str) -> Whois:
-        return Whois.parse_obj(self._query("/whois", refang(entity)))
+        return Whois.model_validate(self._query("/whois", refang(entity)))

--- a/wtfis/clients/shodan.py
+++ b/wtfis/clients/shodan.py
@@ -19,7 +19,7 @@ class ShodanClient:
 
     def get_ip(self, ip: str) -> Optional[ShodanIp]:
         try:
-            return ShodanIp.parse_obj(self.s.host(ip, minify=False))
+            return ShodanIp.model_validate(self.s.host(ip, minify=False))
         except APIError as e:
             if str(e) == "Invalid API key":
                 raise APIError("Invalid Shodan API key")
@@ -38,11 +38,11 @@ class ShodanClient:
             ip_data = self.get_ip(ip.attributes.ip_address)
             if ip_data:
                 shodan_map[ip_data.ip_str] = ip_data
-        return ShodanIpMap(__root__=shodan_map)
+        return ShodanIpMap.model_validate(shodan_map)
 
     def single_get_ip(self, ip: str) -> ShodanIpMap:
         shodan_map = {}
         ip_data = self.get_ip(ip)
         if ip_data:
             shodan_map[ip] = ip_data
-        return ShodanIpMap(__root__=shodan_map)
+        return ShodanIpMap.model_validate(shodan_map)

--- a/wtfis/clients/virustotal.py
+++ b/wtfis/clients/virustotal.py
@@ -26,19 +26,19 @@ class VTClient(BaseClient):
         return "Virustotal"
 
     def get_domain(self, domain: str) -> Domain:
-        return Domain.parse_obj(self._get(f"/domains/{domain}"))
+        return Domain.model_validate(self._get(f"/domains/{domain}"))
 
     def get_domain_resolutions(self, domain: str) -> Resolutions:
-        return Resolutions.parse_obj(self._get(f"/domains/{domain}/resolutions"))
+        return Resolutions.model_validate(self._get(f"/domains/{domain}/resolutions"))
 
     def get_domain_whois(self, domain: str) -> Whois:
-        return Whois.parse_obj(self._get(f"/domains/{domain}/historical_whois"))
+        return Whois.model_validate(self._get(f"/domains/{domain}/historical_whois"))
 
     def get_ip_address(self, ip: str) -> IpAddress:
-        return IpAddress.parse_obj(self._get(f"/ip_addresses/{ip}"))
+        return IpAddress.model_validate(self._get(f"/ip_addresses/{ip}"))
 
     def get_ip_whois(self, ip: str) -> Whois:
-        return Whois.parse_obj(self._get(f"/ip_addresses/{ip}/historical_whois"))
+        return Whois.model_validate(self._get(f"/ip_addresses/{ip}/historical_whois"))
 
     def get_whois(self, entity: str) -> Whois:
         """ Generalized for domain and IP """

--- a/wtfis/models/common.py
+++ b/wtfis/models/common.py
@@ -1,8 +1,14 @@
 import abc
+import sys
 
 from pydantic import BaseModel, BeforeValidator
 from pydantic.v1.validators import str_validator
-from typing import Annotated, List, Optional
+from typing import List, Optional
+
+if sys.version_info >= (3, 9):
+    from typing import Annotated
+else:
+    from typing_extensions import Annotated
 
 
 LaxStr = Annotated[str, BeforeValidator(str_validator)]

--- a/wtfis/models/common.py
+++ b/wtfis/models/common.py
@@ -1,25 +1,29 @@
 import abc
 
-from pydantic import BaseModel
-from typing import List, Optional
+from pydantic import BaseModel, BeforeValidator
+from pydantic.v1.validators import str_validator
+from typing import Annotated, List, Optional
+
+
+LaxStr = Annotated[str, BeforeValidator(str_validator)]
 
 
 class WhoisBase(BaseModel, abc.ABC):
     """ Use to normalize WHOIS fields from different sources """
-    source: Optional[str]
-    domain: Optional[str]
-    registrar: Optional[str]
-    organization: Optional[str]
-    name: Optional[str]
-    email: Optional[str]
-    phone: Optional[str]
-    street: Optional[str]
-    city: Optional[str]
-    state: Optional[str]
-    country: Optional[str]
-    postal_code: Optional[str]
+    source: Optional[str] = None
+    domain: Optional[str] = None
+    registrar: Optional[str] = None
+    organization: Optional[str] = None
+    name: Optional[str] = None
+    email: Optional[str] = None
+    phone: Optional[str] = None
+    street: Optional[str] = None
+    city: Optional[str] = None
+    state: Optional[str] = None
+    country: Optional[str] = None
+    postal_code: Optional[str] = None
     name_servers: List[str] = []
-    date_created: Optional[str]
-    date_changed: Optional[str]
-    date_expires: Optional[str]
-    dnssec: Optional[str]
+    date_created: Optional[str] = None
+    date_changed: Optional[str] = None
+    date_expires: Optional[str] = None
+    dnssec: Optional[str] = None

--- a/wtfis/models/greynoise.py
+++ b/wtfis/models/greynoise.py
@@ -1,6 +1,6 @@
 from __future__ import annotations
 
-from pydantic import BaseModel
+from pydantic import BaseModel, RootModel
 from typing import Dict, Optional
 
 
@@ -10,14 +10,14 @@ class GreynoiseIp(BaseModel):
     riot: bool
     message: str
     link: str
-    classification: Optional[str]
-    name: Optional[str]
-    last_seen: Optional[str]
+    classification: Optional[str] = None
+    name: Optional[str] = None
+    last_seen: Optional[str] = None
 
 
-class GreynoiseIpMap(BaseModel):
-    __root__: Dict[str, GreynoiseIp]
+class GreynoiseIpMap(RootModel):
+    root: Dict[str, GreynoiseIp]
 
     @classmethod
     def empty(cls) -> GreynoiseIpMap:
-        return cls(__root__={})
+        return cls.model_validate({})

--- a/wtfis/models/ip2whois.py
+++ b/wtfis/models/ip2whois.py
@@ -1,4 +1,4 @@
-from pydantic import root_validator, validator
+from pydantic import Field, field_validator, model_validator
 from typing import List, Optional
 
 from wtfis.models.common import WhoisBase
@@ -7,35 +7,25 @@ from wtfis.models.common import WhoisBase
 class Whois(WhoisBase):
     source: str = "ip2whois"
     domain: str = ""
-    registrar: Optional[str]
-    organization: Optional[str]
-    name: Optional[str]
-    email: Optional[str]
-    phone: Optional[str]
-    street: Optional[str]
-    city: Optional[str]
-    state: Optional[str]
-    country: Optional[str]
-    postal_code: Optional[str]
-    name_servers: List[str] = []
-    date_created: Optional[str]
-    date_changed: Optional[str]
-    date_expires: Optional[str]
-    whois_server: Optional[str]
-    dnssec: Optional[str]
+    registrar: Optional[str] = None
+    organization: Optional[str] = None
+    name: Optional[str] = None
+    email: Optional[str] = None
+    phone: Optional[str] = None
+    street: Optional[str] = Field(None, alias="street_address")
+    city: Optional[str] = None
+    state: Optional[str] = Field(None, alias="region")
+    country: Optional[str] = None
+    postal_code: Optional[str] = Field(None, alias="zip_code")
+    name_servers: List[str] = Field([], alias="nameservers")
+    date_created: Optional[str] = Field(None, alias="create_date")
+    date_changed: Optional[str] = Field(None, alias="update_date")
+    date_expires: Optional[str] = Field(None, alias="expire_date")
+    whois_server: Optional[str] = None
+    dnssec: Optional[str] = None
 
-    class Config:
-        fields = {
-            "date_created": "create_date",
-            "date_changed": "update_date",
-            "date_expires": "expire_date",
-            "street": "street_address",
-            "state": "region",
-            "postal_code": "zip_code",
-            "name_servers": "nameservers",
-        }
-
-    @root_validator(pre=True)
+    @model_validator(mode="before")
+    @classmethod
     def extract_registrant(cls, v):
         """ Surface registrant fields to root level """
         registrant = v.pop("registrant", {})
@@ -48,7 +38,8 @@ class Whois(WhoisBase):
             v[field] = registrant.get(field)
         return v
 
-    @validator("registrar", pre=True)
+    @field_validator("registrar", mode="before")
+    @classmethod
     def transform_registrar(cls, v):
         """ Convert registrar from dict to simply registrar.name """
         return v.get("name") if v else v

--- a/wtfis/models/ipwhois.py
+++ b/wtfis/models/ipwhois.py
@@ -1,5 +1,7 @@
-from pydantic import BaseModel
+from pydantic import BaseModel, Field, RootModel
 from typing import Dict
+
+from wtfis.models.common import LaxStr
 
 
 class Flag(BaseModel):
@@ -9,7 +11,7 @@ class Flag(BaseModel):
 
 
 class Connection(BaseModel):
-    asn: str
+    asn: LaxStr
     org: str
     isp: str
     domain: str
@@ -18,7 +20,7 @@ class Connection(BaseModel):
 class IpWhois(BaseModel):
     ip: str
     success: bool
-    type_: str
+    type_: str = Field(alias="type")
     continent: str
     continent_code: str
     country: str
@@ -34,11 +36,6 @@ class IpWhois(BaseModel):
     flag: Flag
     connection: Connection
 
-    class Config:
-        fields = {
-            "type_": "type",
-        }
 
-
-class IpWhoisMap(BaseModel):
-    __root__: Dict[str, IpWhois]
+class IpWhoisMap(RootModel):
+    root: Dict[str, IpWhois]

--- a/wtfis/models/passivetotal.py
+++ b/wtfis/models/passivetotal.py
@@ -1,4 +1,4 @@
-from pydantic import root_validator
+from pydantic import Field, model_validator
 from typing import List, Optional
 
 from wtfis.models.common import WhoisBase
@@ -6,34 +6,24 @@ from wtfis.models.common import WhoisBase
 
 class Whois(WhoisBase):
     source: str = "passivetotal"
-    registrar: Optional[str]
-    organization: Optional[str]
-    name: Optional[str]
-    email: Optional[str]
-    phone: Optional[str]
-    street: Optional[str]
-    city: Optional[str]
-    state: Optional[str]
-    country: Optional[str]
-    postal_code: Optional[str]
-    name_servers: List[str] = []
-    date_created: Optional[str]
-    date_changed: Optional[str]
-    date_expires: Optional[str]
-    dnssec: Optional[str]
+    registrar: Optional[str] = None
+    organization: Optional[str] = None
+    name: Optional[str] = None
+    email: Optional[str] = Field(None, alias="contactEmail")
+    phone: Optional[str] = Field(None, alias="telephone")
+    street: Optional[str] = None
+    city: Optional[str] = None
+    state: Optional[str] = None
+    country: Optional[str] = None
+    postal_code: Optional[str] = Field(None, alias="postalCode")
+    name_servers: List[str] = Field([], alias="nameServers")
+    date_created: Optional[str] = Field(None, alias="registered")
+    date_changed: Optional[str] = Field(None, alias="registryUpdatedAt")
+    date_expires: Optional[str] = Field(None, alias="expiresAt")
+    dnssec: Optional[str] = None
 
-    class Config:
-        fields = {
-            "email": "contactEmail",
-            "phone": "telephone",
-            "name_servers": "nameServers",
-            "postal_code": "postalCode",
-            "date_created": "registered",
-            "date_changed": "registryUpdatedAt",
-            "date_expires": "expiresAt",
-        }
-
-    @root_validator(pre=True)
+    @model_validator(mode="before")
+    @classmethod
     def extract_registrant(cls, v):
         registrant = v.pop("registrant")
         for field in ["telephone", "street", "city", "state", "country", "postalCode"]:

--- a/wtfis/models/shodan.py
+++ b/wtfis/models/shodan.py
@@ -1,28 +1,30 @@
 from collections import defaultdict, namedtuple
-from pydantic import BaseModel
+from pydantic import BaseModel, RootModel
 from typing import Dict, List, Optional
+
+from wtfis.models.common import LaxStr
 
 
 class PortData(BaseModel):
     port: int
-    product: Optional[str]
+    product: Optional[str] = None
     transport: str
 
 
 class ShodanIp(BaseModel):
-    asn: Optional[str]
-    city: Optional[str]
+    asn: Optional[LaxStr] = None
+    city: Optional[str] = None
     country_code: str
     country_name: str
     data: List[PortData]
     ip_str: str
-    isp: Optional[str]
+    isp: Optional[str] = None
     last_update: str
-    org: Optional[str]
-    os: Optional[str]
-    ports: Optional[List[int]]
-    region_name: Optional[str]
-    tags: Optional[List[str]]
+    org: Optional[str] = None
+    os: Optional[str] = None
+    ports: Optional[List[int]] = None
+    region_name: Optional[str] = None
+    tags: Optional[List[str]] = None
 
     def group_ports_by_product(self) -> dict:
         PortProtocol = namedtuple("PortProtocol", "port transport")
@@ -39,5 +41,5 @@ class ShodanIp(BaseModel):
         return result
 
 
-class ShodanIpMap(BaseModel):
-    __root__: Dict[str, ShodanIp]
+class ShodanIpMap(RootModel):
+    root: Dict[str, ShodanIp]

--- a/wtfis/models/virustotal.py
+++ b/wtfis/models/virustotal.py
@@ -1,19 +1,13 @@
-from pydantic import BaseModel, root_validator, validator
+from pydantic import BaseModel, Field, RootModel, field_validator, model_validator
 from typing import Any, Dict, List, Optional
 
-from wtfis.models.common import WhoisBase
+from wtfis.models.common import LaxStr, WhoisBase
 
 
 class BaseData(BaseModel):
-    attributes: Any
-    id_: str
-    type_: str
-
-    class Config:
-        fields = {
-            "id_": "id",
-            "type_": "type",
-        }
+    attributes: Any = None
+    id_: str = Field(alias="id")
+    type_: str = Field(alias="type")
 
 
 class Meta(BaseModel):
@@ -27,8 +21,8 @@ class AnalysisResult(BaseModel):
     result: str
 
 
-class LastAnalysisResults(BaseModel):
-    __root__: Dict[str, AnalysisResult]
+class LastAnalysisResults(RootModel):
+    root: Dict[str, AnalysisResult]
 
 
 class LastAnalysisStats(BaseModel):
@@ -44,29 +38,30 @@ class Popularity(BaseModel):
     timestamp: int
 
 
-class PopularityRanks(BaseModel):
-    __root__: Dict[str, Popularity]
+class PopularityRanks(RootModel):
+    root: Dict[str, Popularity]
 
 
 class BaseAttributes(BaseModel):
-    jarm: Optional[str]
+    jarm: Optional[str] = None
     last_analysis_results: LastAnalysisResults
     last_analysis_stats: LastAnalysisStats
-    last_https_certificate_date: Optional[int]
-    last_modification_date: Optional[int]
+    last_https_certificate_date: Optional[int] = None
+    last_modification_date: Optional[int] = None
     reputation: int
     tags: List[str]
 
 
 class DomainAttributes(BaseAttributes):
     categories: List[str]
-    creation_date: Optional[int]
-    last_dns_records_date: Optional[int]
-    last_update_date: Optional[int]
+    creation_date: Optional[int] = None
+    last_dns_records_date: Optional[int] = None
+    last_update_date: Optional[int] = None
     popularity_ranks: PopularityRanks
-    registrar: Optional[str]
+    registrar: Optional[str] = None
 
-    @validator("categories", pre=True)
+    @field_validator("categories", mode="before")
+    @classmethod
     def transform_categories(cls, v):
         cats = set()
         for category in v.values():
@@ -88,10 +83,10 @@ class Domain(BaseModel):
 
 
 class IpAttributes(BaseAttributes):
-    asn: Optional[int]
-    continent: Optional[str]
-    country: Optional[str]
-    network: Optional[str]
+    asn: Optional[LaxStr] = None
+    continent: Optional[str] = None
+    country: Optional[str] = None
+    network: Optional[str] = None
 
 
 class IpData(BaseData):
@@ -122,38 +117,25 @@ class Resolutions(BaseModel):
 
 class Whois(WhoisBase):
     source: str = "virustotal"
-    domain: str = ""
-    registrar: Optional[str]
-    organization: Optional[str]
-    name: Optional[str]
-    email: Optional[str]
-    phone: Optional[str]
-    street: Optional[str]
-    city: Optional[str]
-    state: Optional[str]
-    country: Optional[str]
-    postal_code: Optional[str]
-    name_servers: List[str] = []
-    date_created: Optional[str]
-    date_changed: Optional[str]
-    date_expires: Optional[str]
-    dnssec: Optional[str]
+    domain: str = Field("", alias="Domain Name")
+    registrar: Optional[str] = None
+    organization: Optional[str] = Field(None, alias="Registrant Organization")
+    name: Optional[str] = None
+    email: Optional[str] = Field(None, alias="Registrant Email")
+    phone: Optional[str] = Field(None, alias="Registrant Phone")
+    street: Optional[str] = Field(None, alias="Registrant Street")
+    city: Optional[str] = Field(None, alias="Registrant City")
+    state: Optional[str] = Field(None, alias="Registrant State/Province")
+    country: Optional[str] = None
+    postal_code: Optional[str] = Field(None, alias="Registrant Postal Code")
+    name_servers: List[str] = Field([], alias="Name Server")
+    date_created: Optional[str] = None
+    date_changed: Optional[LaxStr] = None
+    date_expires: Optional[str] = None
+    dnssec: Optional[str] = Field(None, alias="DNSSEC")
 
-    class Config:
-        fields = {
-            "domain": "Domain Name",
-            "organization": "Registrant Organization",
-            "email": "Registrant Email",
-            "phone": "Registrant Phone",
-            "street": "Registrant Street",
-            "city": "Registrant City",
-            "state": "Registrant State/Province",
-            "postal_code": "Registrant Postal Code",
-            "name_servers": "Name Server",
-            "dnssec": "DNSSEC",
-        }
-
-    @root_validator(pre=True)
+    @model_validator(mode="before")
+    @classmethod
     def get_latest_whois_record_and_transform(cls, v):
         data = v.pop("data")
         if not data:
@@ -195,15 +177,18 @@ class Whois(WhoisBase):
 
         return {**transformed, **fields_w_multiple_possible_sources}
 
-    @validator("name_servers", pre=True)
+    @field_validator("name_servers", mode="before")
+    @classmethod
     def transform_nameservers(cls, v):
         return v.lower().split(" | ")
 
-    @validator("domain", pre=True)
+    @field_validator("domain", mode="before")
+    @classmethod
     def lowercase_domain(cls, v):
         return v.lower()
 
-    @validator("*")
+    @field_validator("*")
+    @classmethod
     def dedupe_values(cls, v):
         if v is not None and "|" in v:
             return v.split(" | ")[0]

--- a/wtfis/ui/base.py
+++ b/wtfis/ui/base.py
@@ -49,7 +49,7 @@ class BaseView(abc.ABC):
 
     def _vendors_who_flagged_malicious(self) -> List[str]:
         vendors = []
-        for key, result in self.entity.data.attributes.last_analysis_results.__root__.items():
+        for key, result in self.entity.data.attributes.last_analysis_results.root.items():
             if result.category == "malicious":
                 vendors.append(key)
         return vendors
@@ -130,16 +130,16 @@ class BaseView(abc.ABC):
         return Text(str(reputation), style=rep_style(reputation))
 
     def _gen_vt_popularity(self, popularity_ranks: PopularityRanks) -> Optional[Text]:
-        if len(popularity_ranks.__root__) == 0:
+        if len(popularity_ranks.root) == 0:
             return None
 
         text = Text()
-        for source, popularity in popularity_ranks.__root__.items():
+        for source, popularity in popularity_ranks.root.items():
             text.append(source, style=self.theme.popularity_source)
             text.append(" (")
             text.append(str(popularity.rank), style=self.theme.inline_stat)
             text.append(")")
-            if source != list(popularity_ranks.__root__.keys())[-1]:
+            if source != list(popularity_ranks.root.keys())[-1]:
                 text.append("\n")
         return text
 
@@ -245,10 +245,10 @@ class BaseView(abc.ABC):
         return text
 
     def _get_ip_enrichment(self, ip: str) -> Optional[Union[IpWhois, ShodanIp]]:
-        return self.ip_enrich.__root__[ip] if ip in self.ip_enrich.__root__.keys() else None
+        return self.ip_enrich.root[ip] if ip in self.ip_enrich.root.keys() else None
 
     def _get_greynoise_enrichment(self, ip: str) -> Optional[GreynoiseIp]:
-        return self.greynoise.__root__[ip] if ip in self.greynoise.__root__.keys() else None
+        return self.greynoise.root[ip] if ip in self.greynoise.root.keys() else None
 
     def whois_panel(self) -> Optional[Panel]:
         # Do nothing if no whois

